### ISSUE TITLE
Record type equality

### ISSUE THREAD

**Issue by zilinc** [NICTA/cogent#205]
<a href="https://github.csiro.au/lim060"><img src="https://github.csiro.au/avatars/u/271?" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [lim060](https://github.csiro.au/lim060)**
_Tue Sep 15 17:50:11 2015_
_Originally opened as https://github.csiro.au/ts-filesystems/Cogent/issues/100_

----

There is a mismatch between `cdsl2` and the formal semantics about record types. To the compiler, field names are significant in type equality for records, but not to the Isabelle/HOL semantics.

This mismatch breaks our C refinement framework because it requires a one-to-one correspondence between CDSL and C types.

There are two obvious fixes where we change either interpretation to match the other. We could also squash record names during compilation, but this could annoy anyone writing C as we'd squash the struct field names as well.

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/che283"><img src="https://github.csiro.au/avatars/u/286?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [che283](https://github.csiro.au/che283)**
_Wed Sep 16 02:04:55 2015_

----

It was changed from a order oblivious implementation to a order sensitive one in the compiler on purpose (c.f. 7956f7f0bb327c916d4dae4227ea8aa76159c90a) to preserve a predictable C interface.

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/lim060"><img src="https://github.csiro.au/avatars/u/271?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lim060](https://github.csiro.au/lim060)**
_Wed Sep 16 11:13:06 2015_

----

Both are order-sensitive, but the Isabelle semantics doesn't have field names. So `{a: U8}` and `{b: U8}` are formally the same type.

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/che283"><img src="https://github.csiro.au/avatars/u/286?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [che283](https://github.csiro.au/che283)**
_Wed Sep 16 11:37:57 2015_

----

that's totally right.. I misunderstood it.

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/tmurray"><img src="https://github.csiro.au/avatars/u/258?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [tmurray](https://github.csiro.au/tmurray)**
_Wed Sep 16 12:20:21 2015_

----

Just to clarify this issue: it would arise only if I had a CDSL2 program that used two structs whose fields respectively had identical types but used different names, right?

Is this a problem for ext2 or bilbyfs currently?

**Comment by zilinc** [comment]
<a href="https://github.csiro.au/lim060"><img src="https://github.csiro.au/avatars/u/271?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lim060](https://github.csiro.au/lim060)**
_Wed Sep 16 21:44:07 2015_

----

The non-uniqueness for C types only rears its head when we try to use the heap refinement. Since DDSL2 doesn't exist, ext2 and BilbyFS don't use the heap very much. The current code only overlaps unboxed types:

```
% cat impl/ext2/build/generated.table | \
  sed -e 's/ReadOnly/Writable/g' -e 's/\(.*\) :=: \(.*\)/\2 \1/' | sort -k2 | \
  { pc=; pcdsl=; while read -r c cdsl; do
    if [[ "$cdsl" = "$pcdsl" && "$c" != "$pc" ]]; then echo "Overlapping: (${cdsl##* }) $c, $pc"; fi;
    pc=$c; pcdsl=$cdsl; done }

Overlapping: (Unboxed) t659, t658
Overlapping: (Unboxed) t445, t196
Overlapping: (Unboxed) t317, t314
Overlapping: (Unboxed) t317, t314
Overlapping: (Unboxed) t689, t373
[...]
```

So it currently doesn't affect either.

**Comment by vjackson725** [comment]
This was fixed in 868a76b75
